pkg/scheduler/backend/api_dispatcher/call_controller.go

```diff
@@ -0,0 +1,300 @@
+//
+// When a new call is added for an object that is already being tracked,
+// the controller reconciles them by merging the calls, overwriting the older one,
+// or skipping the new one entirely based on relevance.
+//
+// The controller tracks the state of all calls (pending, in-flight)
+// to coordinate the asynchronous execution flow managed by the APIDispatcher.
+type callController struct {
+	lock   sync.RWMutex
+	cond   *sync.Cond
+	closed bool
+
+	// apiCallRelevances maps all possible APICallTypes to a relevance value.
+	// A more relevant API call should overwrite a less relevant one for the same object.
+	// Types of the same relevance should only be defined for different object types.
+	apiCallRelevances fwk.APICallRelevances
+	// callIDCounter is a monotonically increasing counter used to assign a unique callID
+	// to each queuedAPICall added to the callController.
+	callIDCounter int
+
+	// apiCalls stores details of API calls for objects that are currently enqueued or in-flight.
+	// Must be used under lock.
+	apiCalls map[types.UID]*queuedAPICall
+	// callsQueue is a FIFO queue that stores the object UIDs of (pending) API calls waiting to be executed.
+	// Must be used under lock.
+	callsQueue buffer.Ring[types.UID]
+	// inFlightEntities stores object UIDs of API calls that are currently in-flight
+	// (i.e., their execution goroutine has been dispatched).
+	// Must be used under lock.
+	inFlightEntities sets.Set[types.UID]
+}
+
+// newCallController returns a new callController object.
+func newCallController(apiCallRelevances fwk.APICallRelevances) *callController {
+	q := callController{
+		apiCallRelevances: apiCallRelevances,
+		apiCalls:          make(map[types.UID]*queuedAPICall),
+		inFlightEntities:  sets.New[types.UID](),
+	}
+	q.cond = sync.NewCond(&q.lock)
+
+	return &q
+}
+
+// isLessRelevant returns true if newCall is less relevant than oldCall.
+func (cc *callController) isLessRelevant(oldCall, newCall fwk.APICallType) bool {
+	return cc.apiCallRelevances[newCall] < cc.apiCallRelevances[oldCall]
+}
+
+// reconcile compares a new API call with an existing one for the same object.
+// Based on their relevance and type, it determines whether to skip the new call,
+// overwrite the old one, or merge them.
+// Must be called under cc.lock.
+func (cc *callController) reconcile(oldAPICall, apiCall *queuedAPICall) error {
+	if cc.isLessRelevant(oldAPICall.CallType(), apiCall.CallType()) {
+		// The new API call is less relevant than the existing one, so skip it.
+		err := fmt.Errorf("a more relevant call is already enqueued for this object: %w", fwk.ErrCallSkipped)
+		apiCall.sendOnFinish(err)
+		return err
+	}
+	if oldAPICall.CallType() != apiCall.CallType() {
+		// API call types don't match, so we overwrite the old one.
+		oldAPICall.sendOnFinish(fmt.Errorf("a more relevant call was enqueued for this object: %w", fwk.ErrCallOverwritten))
+		return nil
+	}
```

Comment:
Questions:
- What does "overwrite" mean here? Looks like, in this case, `oldAPICall` is not removed from the queue, right? So, "putting a new API call while keeping old API call" equals "overwrite" here?
- Can we always overwrite the old one with the new one if the call type is different? Will we force it to users when they define a custom API call?

Reply:
Probably just changing the wording could make it easier to understand. 
So, in short, CallType is the definition for "whether the call is mergeable or not". 
- If the calls should be merged, they should have the same CallType
- If the calls shouldn't be merged, that is, one should overwrite another, they should have different CallTypes.

Also, I still would love to see it documented on `CallType()` func in the interface as well. At least, something like "for how to define CallType, see APICallRelevances's comment".